mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -877,9 +878,10 @@ pub enum S3RequestError {
     #[error("Wrong region (expecting {0})")]
     IncorrectRegion(String),
 
-    /// Forbidden
+    /// Forbidden (metadata is boxed to avoid allocating unnecessary space in case of Result::Ok,
+    /// see "result_large_err" clippy warning for details)
     #[error("Forbidden: {0}")]
-    Forbidden(String),
+    Forbidden(String, Box<ClientErrorMetadata>),
```

Comment:
minor: I'd just simplify this case as well and add optional code and status here

Reply:
Hm, the problem is that for HeadObject we [provide](https://github.com/awslabs/mountpoint-s3/blob/2ec7627756199a0a1426185fa497c013f51cb24c/mountpoint-s3-client/src/s3_crt_client.rs#L1006) "\<no message\>" as the first tuple member, which is not the actual `error_code` from s3 response.

Reporting this in the event log may be misleading, I'd suggest that we keep it as it is now. Another option is to make it `Forbidden(Option<String>, Option<String>, Option<i32>),` and change our existing logging (remove "\<no message\>").

Reply:
I see. I'd be fine with the all `Option` variant, but happy either way.

Reply:
ok, all `Option` lgtm too, the only side effect of this change is logging, Forbidden on HeadObject will now be logged slightly differently:

> 2024-06-18T14:04:07.100018Z  WARN lookup{req=60 ino=132 name="list.txt"}: mountpoint_s3::fuse: lookup failed: inode error: error from ObjectClient: HeadObject failed: Client error: Forbidden: **None**

before it was:
> 2024-06-03T15:57:04.665373Z  WARN lookup{req=50 ino=5 name="list.txt"}: mountpoint_s3::fuse: lookup failed: inode error: error from ObjectClient: HeadObject failed: Client error: Forbidden: **\<no message\>**

Reply:
logging in the case of an existing message will become slightly more verbose

> 2024-06-18T14:17:47.133309Z  WARN read{req=12 ino=2 fh=1 offset=0 size=4096 name="encrypted_with_kms"}: mountpoint_s3::fuse: read failed: get request failed: get object request failed: Client error: Forbidden: **Some("**User: <arn> is not authorized to perform: kms:Decrypt on resource: <arn> because no session policy allows the kms:Decrypt action**")**

the issue is that `Option<String>` does not implement `Display`, so to avoid having `Some` bit in logs, we can do something like this:
```rust
#[derive(Default, Debug)]
pub struct S3RequestErrorCode(Option<String>);

impl Display for S3RequestErrorCode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(error_code) = &self.0 {
            write!(f, "{}", error_code)
        } else {
            write!(f, "<no message>")
        }
    }
}
...
    /// Forbidden. Contains response's error_code, error_message and an http_status.
    #[error("Forbidden: {0}")]
    Forbidden(S3RequestErrorCode, Option<String>, Option<i32>),
...
```

But I don't like this overcomplication. Lets keep `Forbidden(String, ClientErrorMetadata)` (`Box` is not necessary anymore).